apps/xmtp.chat/src/components/Messages/RemoteAttachmentContent.tsx

```diff
@@ -0,0 +1,219 @@
+              return;
+            }
+          }
+        }
+
+        setIsLoading(true);
+        const decryptedAttachment =
+          await RemoteAttachmentCodec.load<Attachment>(content, {
+            codecFor: () => new AttachmentCodec(),
+          });
+
+        const blob = new Blob([decryptedAttachment.data], {
+          type: decryptedAttachment.mimeType,
+        });
+        const blobUrl = URL.createObjectURL(blob);
+
+        // Cache the blob URL
+        urlCache.set(content.url, {
+          blobUrl,
+          failed: false,
+        });
+
+        setDecryptedUrl(blobUrl);
+      } catch {
+        setError("Unable to load attachment");
+        urlCache.set(content.url, {
+          blobUrl: null,
+          failed: true,
+        });
+      } finally {
+        setIsLoading(false);
+        loadingRef.current = false;
+      }
+    },
+    [content],
+  );
+
+  const handleRetry = () => {
+    void loadAttachment(true);
+  };
+
+  // if content changes, reset content display
+  useEffect(() => {
+    setDecryptedUrl(null);
+    setError(null);
+  }, [content]);
+
+  // only load the attachment once on mount
+  // if content changes, users will have to reload the attachment manually
+  useEffect(() => {
+    void loadAttachment();
+  }, []);
+
+  const fileSize = formatFileSize(content.contentLength);
+
+  if (isLoading) {
+    return (
+      <Paper p="sm" radius="md" withBorder>
+        <Group gap="xs" align="center">
+          <Loader size="sm" />
+          <Text size="sm" c="dimmed">
+            Loading attachment...
+          </Text>
+        </Group>
+        <AttachmentDetails
+          filename={content.filename}
+          fileSize={fileSize}
+          align={align}
+        />
+      </Paper>
+    );
+  }
+
+  if (error) {
+    return (
+      <Paper p="sm" radius="md" withBorder>
+        <Box>
+          <Group gap="xs" align="center" mb="xs" wrap="nowrap">
+            <Text size="sm" c="red" style={{ whiteSpace: "nowrap" }}>
+              {error}
+            </Text>
+            <Button size="xs" variant="light" onClick={handleRetry}>
+              Retry
+            </Button>
+          </Group>
+          <AttachmentDetails
+            filename={content.filename}
+            fileSize={fileSize}
+            align={align}
+          />
+        </Box>
+      </Paper>
+    );
+  }
+
+  if (!decryptedUrl) {
+    return (
+      <Paper p="sm" radius="md" withBorder>
+        <Group gap="xs" align="center" mb="xs" wrap="nowrap">
+          <Text size="sm" c="red" style={{ whiteSpace: "nowrap" }}>
+            No content available
+          </Text>
+          <Button size="xs" variant="light" onClick={handleRetry}>
+            Retry
+          </Button>
+        </Group>
+        <AttachmentDetails
+          filename={content.filename}
+          fileSize={fileSize}
+          align={align}
+        />
+      </Paper>
+    );
+  }
+
+  const fileType = getFileType(content.filename);
```

Comment:
Instead of having our own file type recognition, could we do something like `new File(...).type` and use the browser API to determine the file type?

Reply:
we could possibly do this, but it does add in complexity since it would rely on the availability of `decryptedUrl`. also, the `type` field only reads the extension that is passed into `new File(...)`. i think what we have now is fine.